### DISCUSSION TITLE
HTTP secondary cache best practices

### DISCUSSION THREAD

**Discussion by rpopescu** [ccache/ccache#1047] [Questions & answers] [answered]
Hello!
First of all, thanks for the amazing work and the time saved!
Onto the issue: how best to share the cache between a few build machines and users? The outline below doesn't include sharding, but I hope it gets the idea and the question across.
Currently we're using a shared cache on each machine, and I was thinking that the next step is to run an nginx instance on each machine, configured with the /cache location root to the same shared folder that's used and shared by all the users on the machine.
Is there anything wrong with that, and are there better approaches?
A related question - how does redis performance compare with an nginx setup? Any observations welcome.
Many thanks again!

**Comment by rpopescu** [comment]
I've attempted the above on a single machine and set up secondary_cache to itself (so http, nginx, root in the same place as it normally goes in).
Upon not getting a hit in either the primary and secondary, the compilation result is stored in primary in a different folder than in the secondary, and with a slightly different name - in the real example below, file `bbiot2nbvkitq86dai90a5ie8j6hi3o` is placed in (primary)` /var/cache/ccache/9/e/` first, then uploaded to the secondary at `9e/bbiot2nbvkitq86dai90a5ie8j6hi3oM`:
```
[radu@ninja94 ccache]$ md5sum 9/e/bbiot2nbvkitq86dai90a5ie8j6hi3oM 9e/bbiot2nbvkitq86dai90a5ie8j6hi3o 
325eef1941869cf84e04297106b0b639  9/e/bbiot2nbvkitq86dai90a5ie8j6hi3oM
325eef1941869cf84e04297106b0b639  9e/bbiot2nbvkitq86dai90a5ie8j6hi3o
```
The HTTP access log shows that this is indeed what is being sent, so it's not nginx that somehow flattens the hierarchy:
```
"GET /ccache//9e/bbiot2nbvkitq86dai90a5ie8j6hi3o HTTP/1.1" 404 153 "-" "ccache/4.6" "-"
"PUT /ccache//9e/bbiot2nbvkitq86dai90a5ie8j6hi3o HTTP/1.1" 201 0 "-" "ccache/4.6" "-"
```
@jrosdahl I've read the code in HttpStorage.cpp and I see this is intentional. I'm not sure where the "M" at the end of the primary file name goes to, but at least the 9e part is quite clear. Is there an inherent problem in using the exact path and file name as in the primary? Doing that would allow a local compilation result to be used by a remote machine.

**Reply by jrosdahl** [reply]
> Is there an inherent problem in using the exact path and file name as in the primary?

Yes, because the number of directory levels used by primary storage is chosen dynamically. An `abcdef` entry is by default stored as `a/b/cdef`. When the number of files in `a/b` reaches a certain size, the entry will be stored as `a/b/c/def` instead, and if `a/b/c` gets too large it will be stored as `a/b/c/d/ef`. This behavior depends on bookkeeping that is available in primary storage but not in secondary storage, which is just a "dumb" key-value store.

Another problem is that even if the paths were compatible, the bookkeeping for the primary storage would become incorrect if entries were written to a secondary storage that happens to be the same as a primary storage.

**Reply by jrosdahl** [reply]
In https://github.com/ccache/ccache/discussions/1047#discussioncomment-2559024, @rpopescu wrote:

> In short there's no way to share the files between the primary and the secondary currently?

That's correct, and that has never been the goal.

However, what you could do is to have two secondary storages: one on local disk and one over the network. Those can share cache directories.

> If I understand correctly, support for this would require the server side of a secondary storage to use the (same) bookkeeping facilities of (as) the primary storage; for example, a ccache to ccache protocol would be a possible solution, since it would reuse these features automatically.

Something like that, yes, but I don't see that as an interesting use case.

**Comment by rpopescu** [comment]
Thanks for the clarification, @jrosdahl !
In short there's no way to share the files between the primary and the secondary currently?
If I understand correctly, support for this would require the server side of a secondary storage to use the (same) bookkeeping facilities of (as) the primary storage; for example, a ccache to ccache protocol would be a possible solution, since it would reuse these features automatically.

**Reply by jrosdahl** [reply]
Just a tip: You posted a new answer to your original question, but I think you wanted to write a reply to [my reply](https://github.com/ccache/ccache/discussions/1047#discussioncomment-2554518) instead. Something to keep in mind next time. I'm going to reply there to keep the discussion in one thread.

**Comment by jrosdahl** [comment, marked as answer]
> how best to share the cache between a few build machines and users?

Set up and use a shared secondary network storage, possible sharded to spread the load.

> The outline below doesn't include sharding, but I hope it gets the idea and the question across. Currently we're using a shared cache on each machine, and I was thinking that the next step is to run an nginx instance on each machine, configured with the /cache location root to the same shared folder that's used and shared by all the users on the machine. Is there anything wrong with that, and are there better approaches?

I'm not sure I quite understand this idea, I'm afraid. If you have one secondary network storage on each machine, who would use of those multiple secondary storages? Do you mean using something like `secondary_storage = http://build-machine1 http://build-machine2 http://build-machine3`? If so, that would risk being slow since doing many network queries will take time, depending on your network speed of course. I think a shared location would be better.

**Reply by rpopescu** [reply]
> > how best to share the cache between a few build machines and users?
> 
> Set up and use a shared secondary network storage, possible sharded to spread the load.

That's what we did, it's just that it is inherently less efficient in both space and (less so) time compared to being able to build something once and use that in both primary and secondary storages.

> > The outline below doesn't include sharding, but I hope it gets the idea and the question across. Currently we're using a shared cache on each machine, and I was thinking that the next step is to run an nginx instance on each machine, configured with the /cache location root to the same shared folder that's used and shared by all the users on the machine. Is there anything wrong with that, and are there better approaches?
> 
> I'm not sure I quite understand this idea, I'm afraid.

The idea is to use the result of a build both between different users on the same machine (via the primary storage, which I called "a shared cache on each machine"), and between machines (via the secondary storage). As it is now this works well enough, but it requires more storage (and time), since the results need to be pushed to the secondary storage as well.

**Reply by afbjorklund** [reply]
I think is the goal of #1010, but would need a new feature before it works:

* https://github.com/ccache/ccache/issues/1010

There was something similar used, in the old "memcached_only" option.

* https://github.com/ccache/ccache/pull/30

----

Despite some older attempts, most of my setups have used File-on-NFS...

But I would be happy to see some updated benchmarks, using Ccache 4.x ?

Like https://ui.adsabs.harvard.edu/abs/2020EPJWC.24505001M/abstract

> "Fast distributed compilation and testing of large C++ projects"

**Reply by rpopescu** [reply]
Thanks, Anders.

**Reply by jrosdahl** [reply]
@rpopescu wrote:

> The idea is to use the result of a build both between different users on the same machine (via the primary storage, which I called "a shared cache on each machine"), and between machines (via the secondary storage). As it is now this works well enough, but it requires more storage (and time), since the results need to be pushed to the secondary storage as well.

Update: As of ccache 4.7, you can set [`remote_only = true`](https://ccache.dev/manual/4.7.3.html#config_remote_only) and then use [remote file storage](https://ccache.dev/manual/4.7.3.html#_file_storage_backend) _locally_ to share the local cache with a remote HTTP cache from the same machine. More info here: https://github.com/ccache/ccache/discussions/1220#discussioncomment-4127621
